graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/util/DirectedGraph.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.schema.schemafirst.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import net.jcip.annotations.NotThreadSafe;
+
+/** A basic directed graph implementation to perform topological sorts. */
+@NotThreadSafe
+public class DirectedGraph<VertexT> {
+
+  // We need to keep track of the predecessor count. For simplicity, use a map to store it
+  // alongside the vertices.
+  private final Map<VertexT, Integer> vertices;
+  private final Multimap<VertexT, VertexT> adjacencyList;
+  private boolean wasSorted;
+
+  public DirectedGraph(Collection<VertexT> vertices) {
+    this.vertices = Maps.newLinkedHashMapWithExpectedSize(vertices.size());
+    this.adjacencyList = LinkedHashMultimap.create();
+
+    for (VertexT vertex : vertices) {
+      this.vertices.put(vertex, 0);
+    }
+  }
+
+  @VisibleForTesting
+  @SafeVarargs
+  DirectedGraph(VertexT... vertices) {
+    this(Arrays.asList(vertices));
+  }
+
+  /**
+   * this assumes that {@code from} and {@code to} were part of the vertices passed to the
+   * constructor
+   */
+  public void addEdge(VertexT from, VertexT to) {
+    Preconditions.checkArgument(vertices.containsKey(from) && vertices.containsKey(to));
+    adjacencyList.put(from, to);
+    vertices.put(to, vertices.get(to) + 1);
+  }
+
+  /** one-time use only, calling this multiple times on the same graph won't work */
+  public List<VertexT> topologicalSort() {
+    Preconditions.checkState(!wasSorted);
+    wasSorted = true;
+
+    Queue<VertexT> queue = new ArrayDeque<>();
+
+    for (Map.Entry<VertexT, Integer> entry : vertices.entrySet()) {
+      if (entry.getValue() == 0) {
+        queue.add(entry.getKey());
+      }
+    }
+
+    List<VertexT> result = Lists.newArrayList();
+    while (!queue.isEmpty()) {
+      VertexT vertex = queue.remove();
+      result.add(vertex);
+      for (VertexT successor : adjacencyList.get(vertex)) {
+        if (decrementAndGetCount(successor) == 0) {
+          queue.add(successor);
+        }
+      }
+    }
+
+    if (result.size() != vertices.size()) {
+      throw new IllegalArgumentException("failed to perform topological sort, graph has a cycle");
+    }
```

Comment:
Soooo, my graph theory is a little rusty... Can you please add some comments in here for people like me? :)

Reply:
It's [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm), I'll add more explanations to the code.